### ISSUE TITLE
Load locales with webpack

### ISSUE THREAD

**Issue by Jenselme** [aurelia/validation#226]
Hi,

I am trying to use this module with webpack. Sadly, it doesn't work as expected. When I init the plugin with:

```js
aurelia.use
        .standardConfiguration()
        .developmentLogging()
        .plugin('aurelia-validation');
```

I get `validation-locale.js:78 Uncaught (in promise) ReferenceError: System is not defined`.

I tracked the bug and it seems that `System.import` is explicitly used in the code to load the locale files [here](https://github.com/aurelia/validation/blob/8642558efe83b1ea9d8c5fd6b304905fa52fc995/src/validation-locale.js#L58) preventing  the module to work with anything else than SystemJS.

I tried to load the locale manually. To do that, I used this code in `main.js`:
```js
import {bootstrap} from 'aurelia-bootstrapper-webpack';
import {ValidationLocale} from '../node_modules/aurelia-validation/src/validation-locale';
import { data } from '../node_modules/aurelia-validation/src/resources/en-US';

bootstrap(function (aurelia) {
    aurelia.use
        .standardConfiguration()
        .developmentLogging();

    ValidationLocale.Repository.addLocale('en-US', data)

    aurelia.start().then(() => aurelia.setRoot('app', document.body));
});
```

Once I did that, 
```js
console.log(ValidationLocale.Repository.instances.has('en-US'))
```
does respond `true`. However, I still have the following errors:
```
validation-locale.js:78 Uncaught (in promise) ReferenceError: System is not defined
validation-property.js:86 Uncaught (in promise) Error: An exception occurred while trying to load the locale(…)
```

To make it work, I had to modify the modules loaders in webpack and add this line:
```js
{ test: /aurelia-validation\/src\/.*\.js$/, loader: 'babel', query: {stage: 0} },
```

Would it be possible to remove the need for `System.import` or allow the user to load a locale manually?

**Comment by EisenbergEffect** [comment]
@PWKad This is the same issue that was reported the other day. We need to switch this to using the Loader abstraction otherwise it won't work with Webpack, require.js, dojo or any non-system.js loader. I'd put this fix at a pretty high priority.

**Comment by Jenselme** [comment]
It is indeed the same issue than #225 but for webpack.

**Comment by Zensavona** [comment]
yeah this is holding me back from moving to webpack also :)

**Comment by Zensavona** [comment]
Is there an example of this loader being used elsewhere? I'd like to give implementing it a shot cc/ @EisenbergEffect @PWKad

**Comment by EisenbergEffect** [comment]
Yes, there are two implementations. The implementation for SystemJS and RequireJS is called loader-default and the implementation for Webpack is called loader-webpack. There shouldn't be a need to implement a new loader. Rather, for some reason the developer who originally authored the validation library (who is no longer part of the team) neglected to use the loader abstraction and instead hardcoded the SystemJS loader (a big error on his part).

There are examples of the loader's use inside of the i18n library as well as in other places like framework and templating. The i18n library might be the most similar in nature.

**Comment by Zensavona** [comment]
Thanks Rob, I'll take a look this morning

**Comment by Zensavona** [comment]
Ok, so I have a couple more questions for @PWKad or @EisenbergEffect 

I spoke already to @PWKad a bit about the architecture in Gitter, and I have the loader instance available [here](https://github.com/aurelia/validation/blob/master/src/validation-locale.js#L58), but I've hit a wall where I'm actually loading the locale modules. Previously the code was calling `System.import(basePath + localeIdentifier)`, the argument resolving to something like `'./aurelia-validation/resources/en-US'`, but that is no longer working. I've tried a couple of different formats (`en-US`, `./resources/en-US`). What's the correct one? None of the Loader usage I can see elsewhere in the aurelia codebase uses local paths to resolve modules.

Any ideas?

**Comment by EisenbergEffect** [comment]
I'm probably missing some context, but I would think it would not be relative, but something like this "aurelia-validation/resources/en-US".

**Comment by Zensavona** [comment]
@EisenbergEffect hence "the argument resolving to something like './aurelia-validation/resources/en-US'" ;)

My question is: what should the path actually be? To be clear, SystemJS's loader likes paths that look like `'./aurelia-validation/resources/en-US'`

**Comment by EisenbergEffect** [comment]
The loader adds them? That shouldn't happen...that's probably why it isn't working correctly.
I can't think why it would do that.

**Comment by Zensavona** [comment]
@EisenbergEffect Well I'm assuming so, this code

```
// basePath + localeIdentifier === 'aurelia-validation/resources/en-US'
loader.loadModule(basePath + localeIdentifier).then((localeData) => {
  resolve(localeData);
});
```

is producing this error: `Error: Cannot find module './aurelia-validation/resources/en-US'.`

I haven't actually dug into the `loader-webpack` code too far.

**Comment by EisenbergEffect** [comment]
Ok, so my guess is that the webpack loader needs some improvements to handle these scenarios. I don't think this is a validation issue.

**Comment by Zensavona** [comment]
My guess is that [here](https://github.com/aurelia/loader-webpack/blob/master/src/index.js#L78) there needs to be a differentiation between local modules and jspm/npm installed ones.

**Comment by stoffeastrom** [comment]
I'm not sure but it looks like you are trying to load "sub modules" and thus needs to configure webpack to include validation as a [submodule](https://github.com/aurelia/skeleton-navigation/issues/344#issuecomment-198301270). E.g something like

```javascript
plugins: [
    new AureliaWebpackPlugin({
      includeSubModules: [
        { moduleId: "aurelia-validation" }
      ]
    })
```

**Comment by EisenbergEffect** [comment]
@Zensavona Give what @stoffeastrom says a try. This is a recent feature added to the aurelia webpack plugin. You might just need to update the plugin and add the configuration above.

**Comment by Zensavona** [comment]
Thanks @stoffeastrom I had no idea about that.

PR incoming @EisenbergEffect

**Comment by mikeesouth** [comment]
I'm trying to get this to work with Webpack and TypeScript but I can't really figure it out. So first of all I added the following to main.ts:
```typescript
aurelia.use.plugin('aurelia-validation');
```

Secondly, I load the submodule in the aurelia-webpack-plugin:
```js
  plugins: [
    new AureliaWebpackPlugin({
      includeSubModules: [
        { moduleId: "aurelia-validation" }
      ]
    })
```

I have also added this to the ambientTypings section of typings.json and run `typings install`:
```json
"aurelia-validation": "npm:aurelia-validation/dist/commonjs/aurelia-validation.d.ts",
```

I get an error about "System is not defined" (full stack trace at the bottom of my post).

I tried adding this to the loaders in the webpack config but it did not help so I removed it again:
```js
{ test: /aurelia-validation\/src\/.*\.js$/, loader: 'babel', query: {stage: 0} },
```






Full stack trace of "System is not defined" error:
```
System is not defined
    at eval (webpack:///./~/aurelia-validation/dist/commonjs/validation-locale.js?:78:9)
    at ValidationLocaleRepository.load (webpack:///./~/aurelia-validation/dist/commonjs/validation-locale.js?:73:12)
    at ValidationConfig.locale (webpack:///./~/aurelia-validation/dist/commonjs/validation-config.js?:100:58)
    at Object.configure (webpack:///./~/aurelia-validation/dist/commonjs/index.js?:119:42)
    at eval (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:281:34)
From previous event:
    at _loadPlugin (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:279:40)
    at eval (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:272:14)
From previous event:
    at loadPlugin (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:271:73)
    at next (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:528:18)
From previous event:
    at next (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:528:54)
    at eval (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:535:14)
From previous event:
    at FrameworkConfiguration.apply (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:521:42)
    at Aurelia.start (webpack:///./~/aurelia-framework/dist/commonjs/aurelia-framework.js?:150:21)
    at eval (webpack:///./src/main.ts?:24:13)
    at eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:96:12)
    at Array.eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:25:19)
    at eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:86:27)
From previous event:
    at eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:78:44)
From previous event:
    at run (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:73:24)
    at Object.eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:100:1)
    at eval (webpack:///./~/aurelia-bootstrapper-webpack/dist/commonjs/aurelia-bootstrapper-webpack.js?:101:30)
    at Object.<anonymous> (http://localhost:3000/main.bundle.js:1087:2)
    at __webpack_require__ (http://localhost:3000/main.bundle.js:572:30)
    at fn (http://localhost:3000/main.bundle.js:96:20)
    at eval (webpack:///./src/main.ts?:3:38)
    at Object.<anonymous> (http://localhost:3000/main.bundle.js:1081:2)
    at __webpack_require__ (http://localhost:3000/main.bundle.js:572:30)
    at fn (http://localhost:3000/main.bundle.js:96:20)
    at eval (webpack:///multi_main?:3:18)
    at Object.<anonymous> (http://localhost:3000/main.bundle.js:625:2)
    at __webpack_require__ (http://localhost:3000/main.bundle.js:572:30)
    at http://localhost:3000/main.bundle.js:618:37
    at http://localhost:3000/main.bundle.js:621:10
```

**Comment by EisenbergEffect** [comment]
This is an issue with the validation plugin itself. @PWKad We really need to get this fixed. It's coming up every day now.

**Comment by PWKad** [comment]
right this is the issue tracking it, reopening for now.

**Comment by PWKad** [comment]
There does not appear to be any documentation on how to develop against a local copy of a plugin for testing, does anyone know how to do this?  Need to be able to reproduce this issue.

**Comment by wshayes** [comment]
Patrick,

With the project we worked on, I put a build step in to build the plugin as part of the application gulp build and used a soft link ‘ln -s <plugin_dist_dir> <npm folder>’ to make it available to the application.  Just also need to make sure you put a watch on the npm folder files.  That was for jspm - haven’t figured out the best way to do this in web pack, but hopefully something similar will work.

**Comment by Zensavona** [comment]
@PWKad clone the skeleton project next to the dir you have the validation plugin, and depend on it in the skeleton project using `file:...` in your package.json.

Just as side note, it would be really nice to have a set of "tests" which are like a document with examples of how this library should work and be used (like http://aurelia.io/validation/ but actually part of the plugin project)

**Comment by PWKad** [comment]
Yep it would be there was a small sample project but it doesn't work anymore after babel updates.  Present welcome with usage samples

**Comment by jsobell** [comment]
For those trying to get it working in the meantime:
`npm install Zensavona/validation#use-aurelia-loader-instead-of-systemjs`
Thanks @Zensavona

**Comment by PWKad** [comment]
I should be pushing out a fix for now in a bit but I'm still trying to create a repro for this

**Comment by PWKad** [comment]
If anyone can swing by and check out this PR - https://github.com/aurelia/validation/pull/236

**Comment by doktordirk** [comment]
fyi: this broke our aurelia-orm tests. seems we need to fix it to 0.6.6 for now
